Added QR Code feature

anguzz opened the pull request:
Added endpoint which sends a SVG of the link generated as QR code 
 Resolves #140

anguzz committed:
qr-code generator

anguzz committed:
#140

anguzz committed:
Merge branch 'main' into qr-code

anguzz committed:
#140

EthanThatOneKid commented on package.json:
```diff
@@ -27,16 +27,20 @@
     "rrule": "^2.6.9",
-    "rss": "^1.2.2"
+    "rss": "^1.2.2",
+    "svelte-qrcode": "^1.0.0"
```
Based on the rest of the changes in your PR, I can tell we ended up just using the `qrcode` package. So, you can go ahead and run `npm uninstall qr-code-styling svelte-qrcode` since they are unused dependencies.

EthanThatOneKid commented on src/lib/constants/utils.ts:
```diff
@@ -0,0 +1,4 @@
+export function clean(link: string): string {
```
I was feeling a little weird about putting utility functions under a `/constants/` directory, so perhaps we can move some code into a new file structure:

```
src/
├─ lib/
│  ├─ qr/
│  │  ├─ utils.ts
│  │  ├─ utils.test.ts
```

In the proposed file structure, you can put the code for the `clean` function into `/src/lib/qr/utils.ts`, but we can probably be more specific by naming it something like `normalizeLinkName`. Along with this function, we also need to move the `utils.test.ts` file into the proposed `/src/lib/qr/` directory. This way, you can now delete `/src/lib/constants/utils.ts` and `/src/lib/constants/utils.test.ts`.

We can also move the code defining the `genQRCodeSVG` function (from `src/routes/[link=link].svg.ts`) into the proposed `utils.ts` file to organize it better.

```ts
// File: /src/lib/qr/utils.ts

// This function is meant to clean links against data from `$lib/constants/links`.
export function normalizeLinkName(name: string): string {
  return name.replace(/[^a-zA-Z0-9-]/g, '').toLowerCase();
}

export function genQRCodeSvg(url: string): Promise<string> {
  return new Promise((resolve, reject) => {
    QRCode.toString(url, { type: 'svg' }, (err, svg) => {
      if (err) reject(err);
      resolve(svg);
    });
  });
}
```

anguzz replied:
Okay! In the new commit I restructured the functions and re named the clean link.

EthanThatOneKid reviewed (changes requested):
Your PR adequately solves the issue at hand very well! I thought of a minor refactor having to do with me having second thoughts on the `/src/constants/utils.ts` file. I am excited to see this feature launch soon!

anguzz committed:
#450

Refactor and file structure change on functions having to do with QR code feature

anguzz committed:
#450

uninstalled qr code dependencies we did not use specifically qr-code-styling and svelte-qrcode

EthanThatOneKid reviewed (approved):
SHEEESH 😎 This PR looks good to me!
